### PR TITLE
add Format.pp_print_array

### PR DESCRIPTION
**Pull request by redianthus**
Hi,

This PR adds a `pp_print_array` function to the `Format` module.

Doing `pp_print_array ~pp_sep pp_v fmt a` is equivalent to `pp_print_list ~pp_sep pp_v fmt (Array.to_list a)`.

As I often have to convert arrays to lists/sequences to print them or write this function myself, I believe this is a useful function to have in the stdlib. 

If you search on [sherlocode](https://sherlocode.com/?q=pp_print_array%20%5C%7Cfprint_array%5C%7Cpp_print_list%20%5C.%5C*Array.to_list) for `pp_print_array \|fprint_array\|pp_print_list \.\*Array.to_list` you'll find a few examples where an equivalent of this function is used or defined.

This is some post-freeze material but it seems I can't add it to the `post-freeze` milestone (@gasche).

EDIT: I noticed a missing `)` in the `pp_print_list` documentation string and I fixed it in the same commit.

### CODE REVIEW

File: Changes
Comment (dra27) [inline Changes]: The `*` should be a `-` (it's not a breaking change, by our OCaml definition of breaking!)

File: stdlib/format.ml
Comment (dra27) [inline stdlib/format.ml]:
```suggestion
  if len > 0 then begin
```

File: testsuite/tests/lib-format/print_array.ml
Comment (dra27) [inline testsuite/tests/lib-format/print_array.ml]: Minor nit, just for less noisy printf formats later on:
```suggestion
let say s = Printf.printf (s ^^ "\n%!");;
```

File: testsuite/tests/lib-format/print_array.ml
Comment (dra27) [inline testsuite/tests/lib-format/print_array.ml]: With the definition above, this (and the others) can be:

```suggestion
  say "empty";
```

File: testsuite/tests/lib-format/print_array.ml
Comment (dra27) [inline testsuite/tests/lib-format/print_array.ml]:
```suggestion
  test (asprintf "%a" pp_print_intarray [||] = "");
```
(or add a comment for what the flush tests)

File: testsuite/tests/lib-format/print_array.ml
Comment (dra27) [inline testsuite/tests/lib-format/print_array.ml]: It's a nit, but this isn't necessary - if the test fails, the exception would be displayed anyway

**Review by dra27 — COMMENTED** [review]
Seems fine to me, thanks - a tweak in `Changes`, a suggestion in the code, and thoughts on the test

**Comment by hhugo** [comment]
Printing mutable data-structure with `Format` can lead to surprising behavior. The reason is that Format can delay the printing. Mutating the data-structure can change the result if it happens after the call to fprintf and before `Format` has flushed that part of the printing stack.
  
I wonder if this should be mention in the doc comment .

File: testsuite/tests/lib-format/print_array.ml
Comment (redianthus) [reply testsuite/tests/lib-format/print_array.ml]: I simply pasted the one in `print_seq` but OK. :)

**Comment by redianthus** [comment]
@hhugo, I added a note about it in the doc.

**Comment by dbuenzli** [comment]
@zapashcanon it's getting hotter here. I would like however to suggest to rather start by adding: 

```ocaml
Format.pp_print_iter 
 ?pp_sep:(formatter -> unit -> unit) ->
 (('a -> unit) -> 'b -> unit) -> 
 (formatter -> 'a -> unit) -> 
 (formatter -> 'b -> unit)
```
you can find a doc string [here](https://github.com/b0-system/b0/blob/914aeaae0861b185377267ccf1366d030453ae8f/src/b00/std/b00_std.mli#L167-L170) and an implementation [there](https://github.com/b0-system/b0/blob/914aeaae0861b185377267ccf1366d030453ae8f/src/b00/std/b00_std.ml#L154-L160).

You can then easily rewrite the existing `pp_print_list` and the `pp_print_array` you propose with: 

```ocaml
let pp_print_list ?pp_sep pp_elt ppf v =
  pp_print_iter ?pp_sep List.iter pp_elt ppf v

let pp_print_array ?pp_sep pp_elt ppf v =
  pp_print_iter ?pp_sep Array.iter pp_elt ppf v
```
That I hope should make upstream happier since it kills a few locs for them. `pp_print_iter` is also a useful tool for any `iter`able data structure. 

This includes `string` and `bytes` values and with appropriate primitives in `Char` makes it easy to quickly devise a much broader family of hex printers than what is being proposed in https://github.com/ocaml/ocaml/pull/11320

**Comment by redianthus** [comment]
@dbuenzli I'm on it, will update the PR soon.

**Comment by redianthus** [comment]
@dbuenzli it's done. In the changes file I mentioned you as an author as now it's mostly your code. I believe this can be merged.

File: stdlib/format.mli
Comment (dbuenzli) [inline stdlib/format.mli]: Is it really worth having `'b` and `'d` different from `unit` ? 

Suppose one day the `Stdlib` introduces the abbreviation `'a Format.t = Format.formatter -> 'a -> unit` to clarify these long winded formatter signatures. We won't be able to use it here (unless compat is broken).

File: stdlib/format.mli
Comment (dbuenzli) [inline stdlib/format.mli]: This should be `@since 5.1.0`.

File: stdlib/format.mli
Comment (dbuenzli) [inline stdlib/format.mli]: This should be `@since 5.1.0`.

**Review by dbuenzli — COMMENTED** [review]
Thanks @zapashcanon. I have one reservation about the signature of `pp_print_iter` but other than that and two doc nits it looks good to me.

**Comment by redianthus** [comment]
fixed

**Comment by redianthus** [comment]
@dra27 I believe this can be merged ?

**Comment by dbuenzli** [comment]
You need approval from two core developers. Currently there are none.

File: stdlib/format.ml
Comment (nojb) [inline stdlib/format.ml]:
```suggestion
    if !is_first then is_first := false else pp_sep ppf ();
```

File: stdlib/format.ml
Comment (nojb) [inline stdlib/format.ml:1265]: If you are reimplementing `pp_print_list`, there's no reason not to do `pp_print_seq` as well.

File: Changes
Comment (nojb) [inline Changes]: Nitpicking, but normally we don't mention features which are invisible to the user ("rewrite ... using ...").

**Review by nojb — APPROVED** [review]
LGTM

**Comment by redianthus** [comment]
@gasche, in case you have some time to review this one :)

**Review by dra27 — APPROVED** [review]
LGTM, too, thanks!

**Comment by dbuenzli** [comment]
Someone should click the merge button, @zapashcanon has been very patient :-)

**Comment by redianthus** [comment]
Thanks !

**Pull request merged**
